<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform Protein to Protein DIAMOND Searches (BLASTP) — diamond_protein_to_protein • rdiamond</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform Protein to Protein DIAMOND Searches (BLASTP) — diamond_protein_to_protein" />
<meta property="og:description" content="Run protein to protein DIAMOND of reference sequences
against a blast-able database or fasta file." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rdiamond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform Protein to Protein DIAMOND Searches (BLASTP)</h1>
    
    <div class="hidden name"><code>diamond_protein_to_protein.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run protein to protein DIAMOND of reference sequences
against a blast-able database or fasta file.</p>
    </div>

    <pre class="usage"><span class='fu'>diamond_protein_to_protein</span>(
  <span class='kw'>query</span>,
  <span class='kw'>subject</span>,
  output_path = <span class='kw'>NULL</span>,
  is_subject_db = <span class='fl'>FALSE</span>,
  task = <span class='st'>"blastp"</span>,
  sensitivity_mode = <span class='st'>"ultra-sensitive"</span>,
  db_import = <span class='fl'>FALSE</span>,
  postgres_user = <span class='kw'>NULL</span>,
  evalue = <span class='fl'>0.001</span>,
  out_format = <span class='st'>"csv"</span>,
  cores = <span class='fl'>1</span>,
  max_target_seqs = <span class='fl'>500</span>,
  hard_mask = <span class='fl'>TRUE</span>,
  diamond_exec_path = <span class='kw'>NULL</span>,
  add_makedb_options = <span class='kw'>NULL</span>,
  add_diamond_options = <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query</th>
      <td><p>path to input file in fasta format.</p></td>
    </tr>
    <tr>
      <th>subject</th>
      <td><p>path to subject file in fasta format or blast-able database.</p></td>
    </tr>
    <tr>
      <th>output_path</th>
      <td><p>path to folder at which DIAMOND output table shall be stored.
Default is <code>output.path = NULL</code> (hence <code><a href='https://rdrr.io/r/base/getwd.html'>getwd()</a></code> is used).</p></td>
    </tr>
    <tr>
      <th>is_subject_db</th>
      <td><p>logical specifying whether or not the <code>subject</code> file is a file in fasta format (<code>is.subject.db = FALSE</code>; default)
or a <code>fasta</code> file that was previously converted into a blast-able database using <code>makeblastdb</code> (<code>is.subject.db = TRUE</code>).</p></td>
    </tr>
    <tr>
      <th>task</th>
      <td><p>protein search task option. Options are:</p><ul>
<li><p><code>task = "blastp"</code> : Standard protein-protein comparisons (default).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>sensitivity_mode</th>
      <td><p>specify the level of alignment sensitivity. The higher the sensitivity level, the more deep homologs can be found, but at the cost of reduced computational speed.</p><ul>
<li><p><code>sensitivity_mode = "fast"</code> : fastest alignment mode, but least sensitive (default).</p></li>
<li><p><code>sensitivity_mode = "more-sensitive"</code> : second fastest alignment mode, but second least sensitive.</p></li>
<li><p><code>sensitivity_mode = "very-sensitive"</code> : sensitive alignment mode.</p></li>
<li><p><code>sensitivity_mode = "ultra-sensitive"</code> : most sensitive alignment mode (sensitivity as high as BLASTP).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>db_import</th>
      <td><p>shall the DIAMOND hit output be stored in a PostgresSQL database and shall a connection be established to this database? Default is <code>db_import = FALSE</code>.
In case users wish to to only generate a DIAMOND output file without importing it to the current R session they can specify <code>db_import = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>postgres_user</th>
      <td><p>when <code>db_import = TRUE</code> and <code>out_format = "postgres"</code> is selected, the DIAMOND output is imported and stored in a
PostgresSQL database. In that case, users need to have PostgresSQL installed and initialized on their system.
Please consult the Installation Vignette for details.</p></td>
    </tr>
    <tr>
      <th>evalue</th>
      <td><p>Expectation value (E) threshold for saving hits (default: <code>evalue = 0.001</code>).</p></td>
    </tr>
    <tr>
      <th>out_format</th>
      <td><p>a character string specifying the format of the file in which the DIAMOND results shall be stored.
Available options are:</p><ul>
<li><p><code>out_format = "pair"</code> : Pairwise</p></li>
<li><p><code>out_format = "xml"</code> : XML</p></li>
<li><p><code>out_format = "csv"</code> : Comma-separated file</p></li>
</ul></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of cores for parallel BLAST searches.</p></td>
    </tr>
    <tr>
      <th>max_target_seqs</th>
      <td><p>maximum number of aligned sequences that shall be retained. Please be aware that <code>max_target_seqs</code> selects best hits based on the database entry and not by the best e-value. See details here: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty833/5106166 .</p></td>
    </tr>
    <tr>
      <th>hard_mask</th>
      <td><p>shall low complexity regions be hard masked with TANTAN? Default is <code>db_hard_mask = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>diamond_exec_path</th>
      <td><p>a path to the DIAMOND executable or <code>conda/miniconda</code> folder.</p></td>
    </tr>
    <tr>
      <th>add_makedb_options</th>
      <td><p>a character string specifying additional makedb options that shall be passed on to the diamond makedb command line call, e.g. <code>add_make_options = "--taxonnames"</code> (Default is <code>add_diamond_options = NULL</code>).</p></td>
    </tr>
    <tr>
      <th>add_diamond_options</th>
      <td><p>a character string specifying additional diamond options that shall be passed on to the diamond command line call, e.g. <code>add_diamond_options = "--block-size 4.0 --compress 1 --no-self-hits"</code> (Default is <code>add_diamond_options = NULL</code>).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># run diamond assuming that the diamond executable is available</span>
<span class='co'># via the system path ('diamond_exec_path = NULL') and using</span>
<span class='co'># sensitivity_mode = "ultra-sensitive"</span>
<span class='kw'>diamond_example</span> <span class='op'>&lt;-</span> <span class='fu'>diamond_protein_to_protein</span>(
              query   = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/qry_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
              subject = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/sbj_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
              sensitivity_mode = <span class='st'>"ultra-sensitive"</span>,
              output_path = <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(),
              db_import  = <span class='fl'>FALSE</span>)

<span class='co'># look at DIAMOND results</span>
<span class='kw'>diamond_example</span>

<span class='co'># run diamond assuming that the diamond executable is available</span>
<span class='co'># via the miniconda path ('diamond_exec_path = "/opt/miniconda3/bin/"')</span>
<span class='co'># and using 2 cores as well as sensitivity_mode = "ultra-sensitive"</span>
<span class='kw'>diamond_example_conda</span> <span class='op'>&lt;-</span> <span class='fu'>diamond_protein_to_protein</span>(
query   = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/qry_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
subject = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/sbj_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
sensitivity_mode = <span class='st'>"ultra-sensitive"</span>, diamond_exec_path = <span class='st'>"/opt/miniconda3/bin/"</span>,
output_path = <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(),
db_import  = <span class='fl'>FALSE</span>, cores = <span class='fl'>2</span>)

<span class='co'># look at DIAMOND results</span>
<span class='kw'>diamond_example_conda</span>

<span class='co'># run diamond assuming that the diamond executable is available</span>
<span class='co'># via the system path ('diamond_exec_path = NULL') and using</span>
<span class='co'># sensitivity_mode = "ultra-sensitive" and adding command line options:</span>
<span class='co'># "--block-size 4.0 --compress 1 --no-self-hits"</span>
<span class='kw'>diamond_example_ultra_sensitive_add_diamond_options</span> <span class='op'>&lt;-</span> <span class='fu'>diamond_protein_to_protein</span>(
query   = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/qry_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
subject = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/sbj_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
sensitivity_mode = <span class='st'>"ultra-sensitive"</span>,
max_target_seqs = <span class='fl'>500</span>,
output_path = <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(),
db_import  = <span class='fl'>FALSE</span>,
add_diamond_options = <span class='st'>"--block-size 4.0 --compress 1 --no-self-hits"</span>,
cores = <span class='fl'>1</span>
)

<span class='co'># look at DIAMOND results</span>
<span class='kw'>diamond_example_ultra_sensitive_add_diamond_options</span>

<span class='co'># run diamond assuming that the diamond executable is available</span>
<span class='co'># via the system path ('diamond_exec_path = NULL') and using</span>
<span class='co'># sensitivity_mode = "ultra-sensitive" and adding makedb command line options:</span>
<span class='co'># "--taxonnames"</span>
<span class='kw'>diamond_example_ultra_sensitive_add_makedb_options</span> <span class='op'>&lt;-</span> <span class='fu'>diamond_protein_to_protein</span>(
query   = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/qry_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
subject = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/sbj_aa.fa'</span>, package = <span class='st'>'rdiamond'</span>),
sensitivity_mode = <span class='st'>"ultra-sensitive"</span>,
max_target_seqs = <span class='fl'>500</span>,
output_path = <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(),
db_import  = <span class='fl'>FALSE</span>,
add_makedb_options = <span class='st'>"--taxonnames"</span>,
cores = <span class='fl'>1</span>
)

<span class='co'># look at DIAMOND results</span>
<span class='kw'>diamond_example_ultra_sensitive_add_makedb_options</span>
}

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


