<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform Protein to Protein DIAMOND2 Searches (BLASTP) against the NCBI non-redundant (NR) database — diamond_protein_against_ncbi_nr • rdiamond</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform Protein to Protein DIAMOND2 Searches (BLASTP) against the NCBI non-redundant (NR) database — diamond_protein_against_ncbi_nr"><meta property="og:description" content="Run protein to protein DIAMOND2 of reference sequences
against a blast-able database or fasta file."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rdiamond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform Protein to Protein DIAMOND2 Searches (BLASTP) against the NCBI non-redundant (NR) database</h1>
    
    <div class="hidden name"><code>diamond_protein_against_ncbi_nr.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run protein to protein DIAMOND2 of reference sequences
against a blast-able database or fasta file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">diamond_protein_against_ncbi_nr</span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  <span class="va">ncbi_nr_folder_path_path</span>,</span>
<span>  make_diamond_db <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  store_hit_table_in_tmp <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  store_hit_table_at_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  task <span class="op">=</span> <span class="st">"blastp"</span>,</span>
<span>  sensitivity_mode <span class="op">=</span> <span class="st">"fast"</span>,</span>
<span>  use_arrow_duckdb_connection <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  evalue <span class="op">=</span> <span class="fl">1e-20</span>,</span>
<span>  out_format <span class="op">=</span> <span class="st">"csv"</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  max_target_seqs <span class="op">=</span> <span class="st">"unlimited"</span>,</span>
<span>  hard_mask <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  diamond_exec_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_makedb_options <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  add_diamond_options <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>query</dt>
<dd><p>path to input file in fasta format.</p></dd>


<dt>ncbi_nr_folder_path_path</dt>
<dd><p>path to the directory that either stores the raw NCBI NR database
with assumed name <code>nr.gz</code> (requires <code>make_diamond_db = TRUE</code>) or to the already formatted <code>nr_diamond.dmnd</code> or <code>nr</code> database (assumes default: <code>make_diamond_db = FALSE</code>).</p></dd>


<dt>make_diamond_db</dt>
<dd><p>logical specifying whether or not the NCBI NR database at <code>ncbi_nr_folder_path_path</code> should be formatted with <code>diamond makedb</code> (<code>make_diamond_db = FALSE</code>; default).</p></dd>


<dt>task</dt>
<dd><p>protein search task option. Options are:</p><ul><li><p><code>task = "blastp"</code> : Standard protein-protein comparisons (default).</p></li>
</ul></dd>


<dt>sensitivity_mode</dt>
<dd><p>specify the level of alignment sensitivity. The higher the sensitivity level, the more deep homologs can be found, but at the cost of reduced computational speed.</p><ul><li><p><code>sensitivity_mode = "faster"</code> : fastest alignment mode, but least sensitive (default). Designed for finding hits of &gt;70<!-- % identity and short read alignment such as NGS reads. --></p></li>
<li><p><code>sensitivity_mode = "default"</code> : Default mode. Designed for finding hits of &gt;70<!-- % identity and short read alignment such as NGS reads. --></p></li>
<li><p><code>sensitivity_mode = "fast"</code> : fast alignment mode, but least sensitive (default). Designed for finding hits of &gt;70<!-- % identity and short read alignment such as NGS reads. --></p></li>
<li><p><code>sensitivity_mode = "mid-sensitive"</code> : fast alignments between the <code>fast</code> mode and the sensitive mode in sensitivity.</p></li>
<li><p><code>sensitivity_mode = "sensitive"</code> : fast alignments, but full sensitivity for hits &gt;40<!-- % identity. --></p></li>
<li><p><code>sensitivity_mode = "more-sensitive"</code> : more sensitive than the <code>sensitive</code> mode.</p></li>
<li><p><code>sensitivity_mode = "very-sensitive"</code> : sensitive alignment mode.</p></li>
<li><p><code>sensitivity_mode = "ultra-sensitive"</code> : most sensitive alignment mode (sensitivity as high as BLASTP).</p></li>
</ul></dd>


<dt>use_arrow_duckdb_connection</dt>
<dd><p>shall DIAMOND2 hit output table be transformed to an in-process (big data disk-processing) arrow connection to DuckDB? This is useful when the DIAMOND2 output table to too large to fit into memory. Default is <code>use_arrow_duckdb_connection = FALSE</code>.
Please consult the Installation Vignette for details.</p></dd>


<dt>evalue</dt>
<dd><p>Expectation value (E) threshold for saving hits (default: <code>evalue = 0.001</code>).</p></dd>


<dt>out_format</dt>
<dd><p>a character string specifying the format of the file in which the DIAMOND results shall be stored.
Available options are:</p><ul><li><p><code>out_format = "pair"</code> : Pairwise</p></li>
<li><p><code>out_format = "xml"</code> : XML</p></li>
<li><p><code>out_format = "csv"</code> : Comma-separated file</p></li>
</ul></dd>


<dt>cores</dt>
<dd><p>number of cores for parallel DIAMOND searches.</p></dd>


<dt>max_target_seqs</dt>
<dd><p>maximum number of aligned sequences that shall be retained. Please be aware that <code>max_target_seqs</code> selects best hits based on the database entry and not by the best e-value. See details here: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty833/5106166 .</p></dd>


<dt>hard_mask</dt>
<dd><p>shall low complexity regions be hard masked with TANTAN? Default is <code>db_hard_mask = TRUE</code>.</p></dd>


<dt>diamond_exec_path</dt>
<dd><p>a path to the DIAMOND executable or <code>conda/miniconda</code> folder.</p></dd>


<dt>add_makedb_options</dt>
<dd><p>a character string specifying additional makedb options that shall be passed on to the diamond makedb command line call, e.g. <code>add_make_options = "--taxonnames"</code> (Default is <code>add_diamond_options = NULL</code>).</p></dd>


<dt>add_diamond_options</dt>
<dd><p>a character string specifying additional diamond options that shall be passed on to the diamond command line call, e.g. <code>add_diamond_options = "--block-size 4.0 --compress 1 --no-self-hits"</code> (Default is <code>add_diamond_options = NULL</code>).</p></dd>

</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># run diamond assuming that the diamond executable is available</span></span></span>
<span class="r-in"><span><span class="co"># via the system path ('diamond_exec_path = NULL') and using</span></span></span>
<span class="r-in"><span><span class="co"># sensitivity_mode = "ultra-sensitive"</span></span></span>
<span class="r-in"><span><span class="va">diamond_example</span> <span class="op">&lt;-</span> <span class="fu">diamond_protein_against_ncbi_nr</span><span class="op">(</span></span></span>
<span class="r-in"><span>              query   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/qry_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              ncbi_nr_folder_path_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/sbj_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              sensitivity_mode <span class="op">=</span> <span class="st">"ultra-sensitive"</span>,</span></span>
<span class="r-in"><span>              use_arrow_duckdb_connection  <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># look at DIAMOND results</span></span></span>
<span class="r-in"><span><span class="va">diamond_example</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run diamond assuming that the diamond executable is available</span></span></span>
<span class="r-in"><span><span class="co"># via the miniconda path ('diamond_exec_path = "/opt/miniconda3/bin/"')</span></span></span>
<span class="r-in"><span><span class="co"># and using 2 cores as well as sensitivity_mode = "ultra-sensitive"</span></span></span>
<span class="r-in"><span><span class="va">diamond_example_conda</span> <span class="op">&lt;-</span> <span class="fu">diamond_protein_against_ncbi_nr</span><span class="op">(</span></span></span>
<span class="r-in"><span>query   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/qry_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>ncbi_nr_folder_path_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/sbj_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>sensitivity_mode <span class="op">=</span> <span class="st">"ultra-sensitive"</span>, diamond_exec_path <span class="op">=</span> <span class="st">"/opt/miniconda3/bin/"</span>,</span></span>
<span class="r-in"><span>use_arrow_duckdb_connection  <span class="op">=</span> <span class="cn">FALSE</span>, cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># look at DIAMOND results</span></span></span>
<span class="r-in"><span><span class="va">diamond_example_conda</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run diamond assuming that the diamond executable is available</span></span></span>
<span class="r-in"><span><span class="co"># via the system path ('diamond_exec_path = NULL') and using</span></span></span>
<span class="r-in"><span><span class="co"># sensitivity_mode = "ultra-sensitive" and adding command line options:</span></span></span>
<span class="r-in"><span><span class="co"># "--block-size 4.0 --compress 1 --no-self-hits"</span></span></span>
<span class="r-in"><span><span class="va">diamond_example_ultra_sensitive_add_diamond_options</span> <span class="op">&lt;-</span> <span class="fu">diamond_protein_against_ncbi_nr</span><span class="op">(</span></span></span>
<span class="r-in"><span>query   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/qry_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>ncbi_nr_folder_path_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/sbj_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>sensitivity_mode <span class="op">=</span> <span class="st">"ultra-sensitive"</span>,</span></span>
<span class="r-in"><span>max_target_seqs <span class="op">=</span> <span class="fl">500</span>,</span></span>
<span class="r-in"><span>use_arrow_duckdb_connection  <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>add_diamond_options <span class="op">=</span> <span class="st">"--block-size 4.0 --compress 1 --no-self-hits"</span>,</span></span>
<span class="r-in"><span>cores <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># look at DIAMOND results</span></span></span>
<span class="r-in"><span><span class="va">diamond_example_ultra_sensitive_add_diamond_options</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run diamond assuming that the diamond executable is available</span></span></span>
<span class="r-in"><span><span class="co"># via the system path ('diamond_exec_path = NULL') and using</span></span></span>
<span class="r-in"><span><span class="co"># sensitivity_mode = "ultra-sensitive" and adding makedb command line options:</span></span></span>
<span class="r-in"><span><span class="co"># "--taxonnames"</span></span></span>
<span class="r-in"><span><span class="va">diamond_example_ultra_sensitive_add_makedb_options</span> <span class="op">&lt;-</span> <span class="fu">diamond_protein_against_ncbi_nr</span><span class="op">(</span></span></span>
<span class="r-in"><span>query   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/qry_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>ncbi_nr_folder_path_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'seqs/sbj_aa.fa'</span>, package <span class="op">=</span> <span class="st">'rdiamond'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>sensitivity_mode <span class="op">=</span> <span class="st">"ultra-sensitive"</span>,</span></span>
<span class="r-in"><span>max_target_seqs <span class="op">=</span> <span class="fl">500</span>,</span></span>
<span class="r-in"><span>use_arrow_duckdb_connection  <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>add_makedb_options <span class="op">=</span> <span class="st">"--taxonnames"</span>,</span></span>
<span class="r-in"><span>cores <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># look at DIAMOND results</span></span></span>
<span class="r-in"><span><span class="va">diamond_example_ultra_sensitive_add_makedb_options</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

